npm/pkg/dataplane/policies/policymanager_windows.go

```diff
@@ -125,6 +125,98 @@ func (pMgr *PolicyManager) reconcile() {
+			delete(policyKeys, policyKey)
+			continue
+		}
+
+		// 1. remove stale endpoints from policy.PodEndpoints and skip adding to endpoints that already have the policy
+		if policy.PodEndpoints == nil {
+			policy.PodEndpoints = make(map[string]string)
+		}
+
+		epID, ok := policy.PodEndpoints[epToModifyIP]
+		if ok {
+			if epID == epToModifyID {
+				klog.Infof("[PolicyManagerWindows] while adding all policies, will not add policy %s to endpoint since it already exists there. endpoint IP: %s, endpoint ID: %s", policy.PolicyKey, epToModifyIP, epToModifyID)
+				delete(policyKeys, policyKey)
+				continue
+			}
+
+			// If the expected ID is not same as epID, there is a chance that old pod got deleted
+			// and same IP is used by new pod with new endpoint.
+			// so we should delete the non-existent endpoint from policy reference
+			klog.Infof("[PolicyManagerWindows] while adding all policies, removing endpoint from policy's current endpoints since the endpoint ID has changed. policy: %s, endpoint IP: %s, new ID: %s, previous ID: %s", policy.PolicyKey, epToModifyIP, epToModifyID, epID)
+			delete(policy.PodEndpoints, epToModifyIP)
+		}
+
+		// 2. add this policy's rules to a batch
+		policyRules, err := pMgr.getSettingsFromACL(policy)
+		if err != nil {
+			return fmt.Errorf("error while getting settings while applying all policies. err: %w", err)
+		}
+
+		if len(ruleBatches) == 0 {
+			// this is the first NetPol we've seen, so create the first batch of rules
+			ruleBatches = append(ruleBatches, policyRules)
+			continue
+		}
+
+		batch := ruleBatches[len(ruleBatches)-1]
+		if len(batch)+len(policyRules) > pMgr.MaxBatchedACLsPerPod {
+			// create a new batch since adding this NetPol's rules to the batch would exceed the max rules per batch
+			ruleBatches = append(ruleBatches, policyRules)
+		} else {
+			// add this NetPol's rules to the batch
+			batch = append(batch, policyRules...)
+			ruleBatches[len(ruleBatches)-1] = batch
+		}
+	}
+
+	for i, batch := range ruleBatches {
+		klog.Infof("[PolicyManagerWindows] processing batch %d out of %d for adding all policies to endpoint. endpoint ID: %s", i+1, len(ruleBatches), epToModifyID)
+
+		epPolicyRequest, err := getEPPolicyReqFromACLSettings(batch)
+		if err != nil {
+			return fmt.Errorf("error while applying all policies for batch %d out of %d. ruleBatch: %+v. err: %w", i+1, len(ruleBatches), batch, err)
+		}
+
+		klog.Infof("[PolicyManager] applying all rules to endpoint for batch %d out of %d. endpoint ID: %s", i+1, len(ruleBatches), epToModifyID)
+		err = pMgr.applyPoliciesToEndpointID(epToModifyID, epPolicyRequest)
+		if err != nil {
+			return fmt.Errorf("failed to add all policies on endpoint for batch %d out of %d. ruleBatch: %+v. err: %w", i+1, len(ruleBatches), batch, err)
+		}
+
+		klog.Infof("[PolicyManager] finished applying all rules to endpoint for batch %d out of %d. endpoint ID: %s", i+1, len(ruleBatches), epToModifyID)
+		for policyKey := range policyKeys {
```

Comment:
What if you applied a portion of policies and you error out ? you need to persist the state of policies you already applied on that endpoint so we dont end up with duplicate policies?